Standardize expanded TensorOptions representation in native_functions.yaml

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33634 Add MemoryFormat to TensorOptions in codegen.
* #33633 Calls to Tensor::to pass MemoryFormat by TensorOptions
* #33627 Calls to _empty_affine_quantized pass MemoryFormat by TensorOptions
* #33619 Change default value of MemoryFormat? to None
* **#33362 Standardize expanded TensorOptions representation in native_functions.yaml**
* #33518 Collapse _like overloads into a single overload.
* #33516 Add MemoryFormat to TensorOptions, but not codegen.
* #33510 Correctly preserve "not set anywhere" TensorOptions when merging.
* #33505 Switch empty_like to use merge_in to process TensorOptions.

Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I made `pin_memory` nullable as it is more backwards compatible.
  But arguably the right thing to do is make it match Python, e.g.,
  `bool pin_memory=False`

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975409](https://our.internmc.facebook.com/intern/diff/D19975409)

ezyang committed:
Standardize expanded TensorOptions representation in native_functions.yaml

Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@houseroad you are cc'ed because there may be BC implications (though I don't think I broke BC). @VitalyFedyunin you are cc'ed because I modified MemoryFormat overloads.

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

izdeby commented on aten/src/ATen/native/TensorFactories.cpp:
```diff
@@ -190,8 +190,8 @@ AT_FORALL_SCALAR_TYPES_AND3(Bool, Half, BFloat16, DEFINE_CAST_OP)
 
 Tensor empty_like(
     const Tensor& self,
-    c10::optional<c10::MemoryFormat> optional_memory_format) {
```
Wont this change the way code behaves?

ezyang replied:
In fact, it did! But in a non-obvious way, see #28093

VitalyFedyunin replied:
Most of behaviour covered with `test_memory_format_empty_like` and similar tests.

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
This is gonna need https://github.com/pytorch/pytorch/pull/33505 first

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975409](https://our.internmc.facebook.com/intern/diff/D19975409)

[ghstack-poisoned]

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975409](https://our.internmc.facebook.com/intern/diff/D19975409)

[ghstack-poisoned]

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I believe making pin_memory a nullable parameter is the epitome
  of "foolish consistency", and we should abide by how things are
  documented in Python (which matches the above form).

- By doing this, I have transformed some previously mandatory overloads
  of TensorOptions into optional overloads of TensorOptions, in the
  C++ API.  For all _like() factory functions, this makes the TensorOptions
  function ambiguous with the MemoryFormat overload, which is also optional.
  These overloads should never have existed in the first place, so I hack around
  this by just making the parameter on the MemoryFormat overload mandatory, so that I pick the TensorOptions
  overload when no arguments are provided.  There's one minor but obvious fix
  for Python argument parsing when I do this.  TorchScript exported model BC can
  <redacted> <redacted> <redacted>.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975409](https://our.internmc.facebook.com/intern/diff/D19975409)

[ghstack-poisoned]

ezyang committed:
Update on "Standardize expanded TensorOptions representation in native_functions.yaml"


Here is the line of thinking that went into the patch:

- In native_parse.py. we have three different ways you can write the
  expanded form of TensorOptions arguments.  This is dumb, there should
  be exactly one form of TensorOptions that is valid, everything
  else should be disallowed.  So I fiat that all TensorOptions arguments
  must have the form

      *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool pin_memory=False

  and I fix all entries in native_functions.yaml to abide by this specific
  format.

  There was a little bit of hemming and hawing about whether or not
  it should be bool pin_memory=False or bool? pin_memory=None.  In the
  end, I made `pin_memory` nullable as it is more backwards compatible.
  But arguably the right thing to do is make it match Python, e.g.,
  `bool pin_memory=False`

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19975409](https://our.internmc.facebook.com/intern/diff/D19975409)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/native_parse.py:
```diff
@@ -277,8 +264,11 @@ def is_tensor_option(argument):
-                    "Unsupported combination of TensorOptions {}, the only currently supported combinations are {}"\
-                    .format(str(topt_representation), str(supported_topt_arguments))
+                    "Unsupported combination of TensorOptions in {}:\n{}\n\n"\
+                    "The only currently supported combinations are:\n{}"\
```
I think this should now just be "the only currently supported combination is", right? 

...that is if we don't want to make it even sterner, like "TensorOptions must appear in exactly the sequence {}" or something.

I'd call it a nit, but it might be confusing to someone who's new and doesn't really know how to interpret a list of half dozen individual arguments along with an inadvertent plural in the error message.

ezyang replied:
I guess I'll let you handle this one, since this PR also needs to get the BC test script updated too.

bhosmer reviewed (approved):
One semi-nit inline but oh so satisfying
